### PR TITLE
Save Output to Video format

### PR DESCRIPTION
**Pull request by Prakash19921206**
Tested and working fine.  

along with `--demo` ,user need to provide `--saveVideo 1 `  (1 means True, 0 means False which is default )
i'm writing video to the same input resolution provided. 
i'm calling `get_fps(self,frame)` because for cpu it will be around 0.4fps for gpu will be around 30 fps

also i have added namedWindow, so that output window can be resized (full screen or whichever size)

### CODE REVIEW

File: flow [hunk 1]
```diff
@@ -47,7 +48,7 @@ if FLAGS.profile:
 	exit()
 
 if FLAGS.demo:
-	tfnet.camera(FLAGS.demo)
+	tfnet.camera(FLAGS.demo,FLAGS.saveVideo)
```
Comment (thtrieu) [inline flow]: I suggest a space coming after the comma

File: net/help.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def load_old_graph(self, ckpt):
         op = tf.assign(var, plh)
         self.sess.run(op, {plh: val})
 
-def camera(self, file):
+def camera(self, file,SaveVideo):
```
Comment (thtrieu) [inline net/help.py]: I suggest a space coming after the comma

File: net/help.py [hunk 3]
```diff
@@ -71,11 +71,12 @@ def camera(self, file):
     elapsed = int()
     start = timer()
     cv2.namedWindow('', 0)
-    fourcc = cv2.VideoWriter_fourcc(*'XVID')
-    _, frame = camera.read()
-    fps = get_fps(self,frame)
-    height, width, _ = frame.shape
-    videoWriter = cv2.VideoWriter('video.avi', fourcc, fps, (width,height))
+    if(SaveVideo):
```
Comment (thtrieu) [inline net/help.py]: `if SaveVideo:`

File: net/help.py [hunk 4]
```diff
@@ -71,11 +71,12 @@ def camera(self, file):
...
+        _, frame = camera.read()
+        fourcc = cv2.VideoWriter_fourcc(*'XVID')
+        fps = get_fps(self,frame)
+        height, width, _ = frame.shape
+        videoWriter = cv2.VideoWriter('video.avi', fourcc, fps, (width,height))
```
Comment (thtrieu) [inline net/help.py]: `(width, height)`

File: net/help.py [hunk 5]
```diff
@@ -71,11 +71,12 @@ def camera(self, file):
...
     while camera.isOpened():
         _, frame = camera.read()
         if (frame is None):
```
Comment (thtrieu) [inline net/help.py]: if frame is None:

File: net/help.py [hunk 6]
```diff
@@ -85,7 +86,8 @@ def camera(self, file):
         feed_dict = {self.inp: [preprocessed]}
         net_out = self.sess.run(self.out,feed_dict)[0]
         processed = self.framework.postprocess(net_out, frame, False)
-        videoWriter.write(processed)
+        if (SaveVideo):
```
Comment (thtrieu) [inline net/help.py]: if SaveVideo:

File: net/help.py [hunk 7]
```diff
@@ -71,11 +71,12 @@ def camera(self, file):
...
     cv2.namedWindow('', 0)
-    fourcc = cv2.VideoWriter_fourcc(*'XVID')
-    _, frame = camera.read()
-    fps = get_fps(self,frame)
-    height, width, _ = frame.shape
-    videoWriter = cv2.VideoWriter('video.avi', fourcc, fps, (width,height))
+    if(SaveVideo):
+        _, frame = camera.read()
+        fourcc = cv2.VideoWriter_fourcc(*'XVID')
+        fps = get_fps(self,frame)
```
Comment (thtrieu) [inline net/help.py]: `fps = get_fps(self, frame)`

**Comment by thtrieu** [comment]
Thanks!

I suggest:
1. A few style fixes as I commented in the code.
2. Adding a section into README, note that `--saveVideo` is enough for saving the video, and not specifying `--saveVideo` means no saving, there is no need for values `0`, `1`.

File: net/help.py (hunk 5 above)
Comment (Prakash19921206) [inline net/help.py]: not mine, will change it anyway :)

**Comment by borasy** [comment]
I tested on an mpeg4 codec video, and i get the following error, and the video is not saved. 

[mpeg4 @ 0x3714460] framerate not set
Could not open codec 'mpeg4': Unspecified error

If i set the fps value to 25, manually, the video is saved fine. @Prakash19921206 @thtrieu

**Comment by thtrieu** [comment]
I am no expert on codec, so waiting for people to come help ...

**Comment by borasy** [comment]
@thtrieu the problem is not with codec here.. the problem is with the fps.. like i said, i put 25 manually in the `videoWriter = cv2.VideoWriter('video.avi', fourcc, 25, (width, height))` and it works...

**Comment by Prakash19921206** [comment]
@borasy,  MPEG4 or mp4 is only container name (aka extension), can you mention codec name?
 use [MediaInfo ](https://sourceforge.net/projects/mediainfo/)software to check details about your video..can you revert back with details shown? i'll try video with same settings
this could be problem of codec although it looks like setting frame rate issue

make sure u have installed codec pack from [here](https://www.google.co.in/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwiJuLCThYXTAhWCVrwKHaJMDrsQFggbMAA&url=http%3A%2F%2Ffilehippo.com%2Fdownload_klite_codec_pack%2F&usg=AFQjCNFrofKhXSn46AvVitG9A9Ko-ERD-g) or [here](http://download.cnet.com/Windows-Media-Codec-Pack/3000-13632_4-75222206.html)
i tested on few video formats like `MPEG-4 Visual (H.264 Codec)` , `MPEG-4 (H.263 Codec)`,  `WMV (WVC1 Codec)` it worked fine

for camera demo, we have to set the same fps as you get output (Note: for camera if you set fps as 25 & record video at lower resolution like 0.5, then playback will be extremely fast)
but for input video we can create smooth playback by getting original fps and setting the same fps in code.
this seems to be a small bug in OpenCv, when time permits today. i'll change code accordingly and create a pull request.

**Comment by borasy** [comment]
@Prakash19921206 yes, i understand what you mean. At this point we can declare that the codec is **not** the problem. Sorry my first comment made this confusion. The problem is with the FPS indeed. I think the simplest solution is to just get the FPS from the input video and set it into VideoWriter(), there's no need of calculation. I used this `fps = camera.get(cv2.CAP_PROP_FPS)` to get the FPS from the original input video, and everything works fine. I can create the pull request if you don't mind.

**Comment by Prakash19921206** [comment]
@borasy , sure please create PR, i'm slightly occupied now.. 
also can you test if the code `fps = camera.get(cv2.CAP_PROP_FPS) ` is working fine for webcam? (check playback once after saving with webcam input)

i think, for webcam we have to calculate fps.. or else it video playback might be faster or slower..
if above happens then , can you find a workaround? there are many stackoverflow questions on this..

also change [this](https://github.com/thtrieu/darkflow/blob/master/net/help.py#L102) line to 
```
    if SaveVideo:
        videoWriter.release()
```
this will avoid one error message, in case if user do not want to save video...

**Comment by borasy** [comment]
My webcam is broken at the moment. After i got a new one, i'll test on the webcam, too. I also fixed your mentioned bug and created a PR already. Have a nice weekend.

**Comment by Prakash19921206** [comment]
`cv2.CAP_PROP_FPS` is not stable. with my desktop, i have usb webcam
when i tried `fps = camera.get(cv2.CAP_PROP_FPS)`  i'm getting 0 for fps hence video also not recording 

but it works for video input

**Comment by borasy** [comment]
@Prakash19921206  I see the problem. This should fix it. 
https://www.learnopencv.com/how-to-find-frame-rate-or-frames-per-second-fps-in-opencv-python-cpp/
But i don't have a webcam to test it yet, so can u fix this one?

**Comment by Prakash19921206** [comment]
i think for webcam, the method i have given to calculate fps is proper (what you see is what you record)

i tried above method, it showed 13fps
so in code i set to 13, for this video playback was very fast. 

only change we need to do is to set `fps = camera.get(cv2.CAP_PROP_FPS)` for video input and not for camera input

**Comment by borasy** [comment]
hmm.. i see. You're right. We should use your get_fps() for camera/webcam. But we also need one more change. Your method is calculating the X time it takes for 1 frame to be processed, right? So if we want to know how many frames it can process in 1 second. then fps should be equals to 1/X. Then it should be
 `X = get_fps(self, frame)` and `fps = 1/X`. Do you agree?

**Comment by Prakash19921206** [comment]
yeah i had missed it..thanks for pointing it out.. what i noticed is, opencv had a bug.
for camera, if input is below 1 fps it doesn't save video....so in code i'm checking if fps is lower than 1
```
if fps < 1:
    fps = 1
```
there was code conflict in PR, after that 1 function was missed from my fork!
waiting for @thtrieu to add `get_fps `in code as said [here ](https://github.com/thtrieu/darkflow/pull/132)

**Comment by borasy** [comment]
Cool. for now that's the way to do it. Awesome job.

**Comment by evgkarasev** [comment]
Well done, guys! It works like a charm!

**Pull request merged**